### PR TITLE
feat: CMS-driven custom My Account pages

### PR DESCRIPTION
**Pull request by lemagnetic**
## What's the purpose of this pull request?

After the My Account CMS migration, **new** account pages created by a store were still code-only — merchants could not author/edit them in CMS Admin. This PR closes that gap ([SFS-3250](https://vtex-dev.atlassian.net/jira/software/c/projects/SFS/boards/1051?selectedIssue=SFS-3250)): stores can declare a My Account route bound to a CMS content-type and render it SSR like native pages.

## How it works?

- Opt-in: optional `contentType` on `Route` in `navigation.ts` (no separate `source` flag).
- **Mode A:** `contentType` + no `.tsx` → body 100% from CMS (CLI generates the page).
- **Mode B:** `contentType` + `.tsx` → CMS sections + code body (+ optional before/after).
- No `contentType` → legacy code-only wrap (unchanged).
- **core:** widened `MyAccountContentType`; new `myAccountCmsServerSideProps` (same auth as native pages).
- **cli:** parses `navigation.ts`, generates pages via `myAccountCmsPageTemplate`; skips native collisions and routes outside `/pvt/account`.

## How to test it?

- Unit tests in `@faststore/core` and `@faststore/cli` (CMS SSR, routes, page generation).
- In a store (e.g. b2bfaststoredev): add `contentType` on a route, add content-type under `cms/faststore/pages/`, `cms-sync`, publish sections in Admin, open the route logged in.
- Regression: routes without `contentType` and existing before/after still work.
- See also: `specs/007-myaccount-cms-custom-pages/quickstart.md`.

## References

- [SFS-3250](https://vtex-dev.atlassian.net/jira/software/c/projects/SFS/boards/1051?selectedIssue=SFS-3250)
- Spec/RFC: `specs/007-myaccount-cms-custom-pages/`

[SFS-3250]: https://vtex-dev.atlassian.net/browse/SFS-3250?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SFS-3250]: https://vtex-dev.atlassian.net/browse/SFS-3250?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by CodeRabbit

* **New Features**
  * Added CMS-driven “My Account” page generation for routes under `/pvt/account` using an optional content type.
  * Supports hybrid pages with customization code and optional before/after sections.
  * Supports custom CMS content types and preserves existing native pages.
  * Added routing safeguards to prevent invalid or conflicting page generation.

* **Tests**
  * Expanded coverage for CMS route parsing, page generation, templates, redirects, and server-side data loading.

### CODE REVIEW

**Review by renatomaurovtex — COMMENTED** [review]
Reviewed the full diff (cli + core) and cross-checked the integration points on the head branch (`fetchMyAccountPageContent`, native `myAccountServerSideProps`, `ACCOUNT_COMPONENTS`, `RenderSectionsBase`, `Layout`). The design is solid: the SSR factory mirrors the native guards exactly (login redirect, 403+`from`, My Account toggle, locale validation), `ACCOUNT_COMPONENTS` already merges GLOBAL + PLUGINS + CUSTOM components so store sections resolve on custom pages, CMS fetch fails open to empty sections, and the CLI enforces the `/pvt/account` prefix and native-page precedence. Test coverage on both packages is good. Findings below are hardening-level — no blockers.

### `packages/cli/src/utils/readMyAccountCmsRoutes.ts`

🟡 **[bug/edge case] Block comments are not stripped — commented-out routes still generate pages**

Only `//` line comments are removed. A route entry disabled with `/* ... */` inside the `routes: [...]` array still matches the object-literal regex, so the CLI generates a page for a route the developer thinks is off:

```ts
routes: [
  /* { route: '/pvt/account/old', title: 'Old', contentType: 'myAccountOld' }, */
]
```

```diff
-  // Strip line comments to avoid false matches inside comments.
-  const withoutLineComments = sourceText.replace(/^\s*\/\/.*$/gm, '')
+  // Strip block and line comments to avoid false matches inside comments.
+  const withoutLineComments = sourceText
+    .replace(/\/\*[\s\S]*?\*\//g, '')
+    .replace(/^\s*\/\/.*$/gm, '')
```

Related, lower priority: `routes\s*:\s*\[([\s\S]*?)\]` stops at the *first* `]`, so if a route entry ever grows an array-valued prop, every entry after it is silently dropped (no warning, unlike the documented dynamic-entry case). Fine for today's `Route` shape — maybe leave a comment on the regex so the next person extending `Route` knows. ⚡

🟡 **[robustness] Validate `contentType`/`route` before interpolating them into generated source**

`myAccountCmsPageTemplate` splices these strings raw between single quotes (`myAccountCmsServerSideProps('${contentType}', '${routePath}')`). A stray `'` in `navigation.ts` (typo, smart quote) produces a syntactically broken generated page and a cryptic Next.js build error far from the cause. Since this is store-authored input parsed by regex, a cheap allowlist here keeps garbage out of codegen: ⚡

```diff
     const contentType = readStringProp(literal, 'contentType')
     if (!contentType) {
       continue
     }
+
+    if (!/^[A-Za-z0-9_-]+$/.test(contentType)) {
+      logger.warn(
+        `[my-account-cms] Skipping route with invalid contentType "${contentType}" in ${navigationFilePath}`
+      )
+      continue
+    }
```

### `packages/core/src/server/cms/myAccountDefaultSections.ts`

🟡 **[bug/edge case] Prototype keys crash the lookup now that `contentType` is an open string**

`DEFAULT_SECTION_KEYS[contentType as NativeMyAccountContentType]` with a store-controlled string means `contentType: 'toString'` (or any `Object.prototype` key) returns a function — the `?? []` never kicks in and `.map` throws, turning a bad nav entry into an SSR 500. Absurd input, one-line guard: ⚡

```diff
-  return (
-    DEFAULT_SECTION_KEYS[contentType as NativeMyAccountContentType] ?? []
-  ).map((key) => ({
+  return (
+    (Object.hasOwn(DEFAULT_SECTION_KEYS, contentType)
+      ? DEFAULT_SECTION_KEYS[contentType as NativeMyAccountContentType]
+      : []) ?? []
+  ).map((key) => ({
```

The `NativeMyAccountContentType | (string & {})` split itself is a nice touch — autocomplete preserved, no breaking change.

### `packages/cli/src/utils/createNextjsPages.ts`

💬 The `if (items.length === 0) return true` early-return is a no-op — `[].every()` already returns `true`. Harmless; keep it only if you want the "Mode A only" intent documented, otherwise drop.

💬 Shadowed-param rename (`basePath` → `pagesBase`) inside `allPagesAreAllowed` is a good catch.

### `packages/cli/src/utils/readMyAccountCmsRoutes.test.ts`

💬 The trailing top-level `afterEach(() => { /* no-op placeholder for symmetry */ })` is dead code — remove it.

### `packages/core/src/experimental/myAccountCmsServerSideProps.ts`

💬 A typo'd or never-published `contentType` renders an *empty* page body with only a server-side `console.warn` — hard for a merchant to self-diagnose. A follow-up dev-mode hint (or a note in the docs/quickstart) would help. Not blocking.

💬 This factory duplicates ~80 lines of guard logic from `myAccountServerSideProps`, and the two have already drifted slightly (this one optional-chains `account.data?.accountProfile?.name ?? ''`; native dereferences unguarded — yours is the better behavior). Consider extracting a shared guard helper later so fixes land in both.

---

## Verdict: Approved with comments

Blocking (🔴/🟠):
- None.

Non-blocking (🟡/💬):
- Strip block comments in `readMyAccountCmsRoutes` (commented-out routes still generate pages).
- Allowlist-validate `contentType` before code generation.
- `Object.hasOwn` guard in `getDefaultMyAccountSections` (prototype-key SSR crash).
- Dead code: redundant empty-dir early return; stray top-level `afterEach` in test.
- Follow-ups: dev-visible signal for unpublished/typo'd content-types; extract shared SSR guard to prevent drift.

Checks to confirm before merge: pnpm lint · build · `pnpm turbo run test --filter=@faststore/cli --filter=@faststore/core` · pnpm size (touches core). SonarQube red is the usual new-code gate; UI Tests (Chromatic) still pending.

File: packages/cli/src/utils/createNextjsPages.test.ts [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+    const generated = path.join(corePagesDir, 'pvt', 'account', 'legacy.tsx')
+    expect(fs.existsSync(generated)).toBe(true)
+    const content = fs.readFileSync(generated, 'utf8')
+    expect(content).toContain('myAccountServerSideProps')
+    expect(content).toContain('DynamicPage')
+    expect(content).not.toContain('myAccountCmsServerSideProps')
+  })
+
+  it('skips CMS generation when the route collides with a native page', () => {
```
Comment (eduardoformiga) [inline packages/cli/src/utils/createNextjsPages.test.ts:120]: nice!

**Review by eduardoformiga — COMMENTED** [review]
do you happen to test it in any store/account? I can't see the content types in the `b2bfaststoredev`. 
Also would be good to ask the education team to create docs about the behaviour and all the possibilities/priorities etc.

**Review by eduardoformiga — APPROVED** [review]

File: packages/cli/src/utils/createNextjsPages.test.ts [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+      'src',
+      'customizations',
+      'src',
+      'myAccount'
+    )
+
+    fs.mkdirSync(corePagesDir, { recursive: true })
+    fs.mkdirSync(customizationPagesDir, { recursive: true })
+    fs.mkdirSync(navigationDir, { recursive: true })
+  })
```
[the remark this answers is not in the document]
Comment (lemagnetic) [reply packages/cli/src/utils/createNextjsPages.test.ts:53]: Not applying this one, on purpose.

`packages/cli/AGENTS.md` says *command* tests **should** mock the file system. These two suites are utility tests whose entire subject **is** filesystem orchestration: directory walking, native-page collision detection, `before/after` extension discovery and page emission. Mocking `node:fs` here would mean asserting against the mock's bookkeeping instead of the behaviour we care about, and it would not catch path/normalization bugs (which is exactly what the `/pvt/accounting` finding in your later review turned out to be).

It is also the established convention in this package — `generate.test.ts`, `cp-schema.test.ts`, `commands/cms-sync.test.ts` and `commands/generate-i18n.test.ts` all create isolated temp dirs. Both suites use `fs.mkdtempSync(os.tmpdir())`, touch nothing outside it, clean up in `afterEach`, and spawn no external processes, so they stay hermetic and parallel-safe.

Happy to revisit if we introduce an injectable fs adapter for the CLI as a whole, but that is a package-wide refactor rather than something for this PR.

File: packages/core/src/experimental/myAccountCmsServerSideProps.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+      }
+    }
+
+    const { isFaststoreMyAccountEnabled, redirect } = getMyAccountRedirect({
+      query: context.query,
+    })
+
+    if (!isFaststoreMyAccountEnabled) {
+      return { redirect }
+    }
```
[the remark this answers is not in the document]
Comment (lemagnetic) [reply packages/core/src/experimental/myAccountCmsServerSideProps.ts:88]: Fixed at the root in 3bc8133.

Runtime-wise `redirect` could never be `null` in this branch (`getMyAccountRedirect` returns a non-null redirect exactly when the flag is off), and `packages/core` compiles with `strict: false`, so nothing was actually broken. But the type was indeed misleading, and the same `return { redirect }` pattern is repeated in all native My Account pages plus `myAccountServerSideProps`.

So I took your second suggestion instead of narrowing locally: `getMyAccountRedirect` now returns a discriminated union, which makes `redirect` narrow to non-null in the `!isFaststoreMyAccountEnabled` branch for every call site at once.

```ts
export type MyAccountRedirectResult =
  | { isFaststoreMyAccountEnabled: false; redirect: MyAccountRedirect }
  | { isFaststoreMyAccountEnabled: true; redirect: null }
```

`tsc --noEmit` on `@faststore/core` is clean.

**Comment by lemagnetic** [comment]
Addressed the outside-diff finding from the last CodeRabbit review (`/pvt/account` prefix boundary) in 92e879a: `isAllowedPrefixPage` now accepts only the exact prefix or `${prefix}/` descendants, so `/pvt/accounting` is rejected. Added a regression test (`does not generate CMS pages for routes that only look like the allowed prefix`) plus a `/pvt/accounting/dashboard` entry in the collision fixture.

The nullable-redirect finding is answered in its thread (fixed via a discriminated union in `getMyAccountRedirect`, 3bc8133). The filesystem-mocking finding I'm declining, with the reasoning in its thread.
